? search.tcl
? tcl/test
Index: search.info
===================================================================
RCS file: /cvsroot/openacs-4/packages/search/search.info,v
retrieving revision 1.46.2.8
diff -u -r1.46.2.8 search.info
--- search.info	1 Mar 2022 09:18:44 -0000	1.46.2.8
+++ search.info	27 Jul 2023 09:15:37 -0000
@@ -36,6 +36,7 @@
             <parameter scope="instance" datatype="string"  min_n_values="1"  max_n_values="1"  name="Symbol2Interval"  default="m3 {3 month ago} m6 {6 month ago} y1 {1 year ago}" description="Translate symbol to date interval."/>
             <parameter scope="instance" datatype="number"  min_n_values="1"  max_n_values="1"  name="Syndicate"  default="0" description="Should syndication data be stored when indexing for search."/>
             <parameter scope="instance" datatype="number"  min_n_values="1"  max_n_values="1"  name="ValidateCSRFP"  default="1" description="Boolean parameter to activate/deactivate CSRF protection for this package instance"/>
+            <parameter scope="instance" datatype="string"  min_n_values="1"  max_n_values="1"  name="VectorDriver"  description="Vector Similarity Search via an FtsEngineDriver implementation. Name the implementation."/>
         </parameters>
 
     </version>
Index: tcl/search-procs.tcl
===================================================================
RCS file: /cvsroot/openacs-4/packages/search/tcl/search-procs.tcl,v
retrieving revision 1.55.2.8
diff -u -r1.55.2.8 search-procs.tcl
--- tcl/search-procs.tcl	1 Mar 2022 09:18:44 -0000	1.55.2.8
+++ tcl/search-procs.tcl	27 Jul 2023 09:15:37 -0000
@@ -1,7 +1,7 @@
 ad_library {
     full-text search engine
 
-    @author Neophytos Demetriou (k2pts@yahoo.com)
+    @author Neophytos Demetriou (neophytos@gmail.com)
     @cvs-id $Id: search-procs.tcl,v 1.55.2.8 2022/03/01 09:18:44 gustafn Exp $
 }
 
@@ -99,22 +99,36 @@
     table and calls the appropriate driver functions to index, update, or
     delete the entry.
 
-    @author Neophytos Demetriou
+    @author Neophytos Demetriou (neophytos@gmail.com)
     @author Jeff Davis (davis@xarg.net)
 } {
 
-    set driver [parameter::get \
-                    -package_id [apm_package_id_from_key search] \
-                    -parameter FtsEngineDriver]
-
-    if { $driver eq ""
-         || (![callback::impl_exists -callback search::index -impl $driver] \
-                 && ! [acs_sc_binding_exists_p FtsEngineDriver $driver])
-     } {
-        # Nothing to do if no driver
-        ns_log Debug "search::indexer: driver=$driver binding exists? " \
-            "[acs_sc_binding_exists_p FtsEngineDriver $driver]"
-        return
+    #set drivers [parameter::get \
+    #                -package_id [apm_package_id_from_key search] \
+    #                -parameter FtsEngineDriver]
+
+    set id [db_string -cache_key fts_engine_driver_contract_id get_fts_engine_contract_id "select acs_sc_contract__get_id('FtsEngineDriver')"]
+    set drivers [db_list -cache_key fts_engine_driver_valid_bindings get_all_fts_valid_bindings {
+	select
+            acs_sc_impl__get_name(b.impl_id) as impl_name
+        from
+            acs_sc_bindings b,
+            acs_sc_impls impl
+        where
+            b.contract_id = :id
+        and impl.impl_id = b.impl_id
+    }]
+    ns_log notice "search indexer drivers=$drivers"
+    foreach driver $drivers {
+	    if { $driver eq ""
+		 || (![callback::impl_exists -callback search::index -impl $driver] \
+			 && ! [acs_sc_binding_exists_p FtsEngineDriver $driver])
+	     } {
+		# Nothing to do if no driver
+		ns_log Debug "search::indexer: driver=$driver binding exists? " \
+		    "[acs_sc_binding_exists_p FtsEngineDriver $driver]"
+		return
+	    }
     }
     # JCD: pull out the rows all at once so we release the handle
     foreach row [db_list_of_lists search_observer_queue_entry {}] {
@@ -164,34 +178,37 @@
                             search::content_get txt $datasource(content) $datasource(mime) \
                                 $datasource(storage_type) $object_id
 
-                            if {[callback::impl_exists -callback search::index -impl $driver]} {
-                                if {![info exists datasource(package_id)]} {
-                                    set datasource(package_id) ""
-                                }
-
-                                if {![info exists datasource(relevant_date)]} {
-                                    set datasource(relevant_date) ""
-                                }
-                                #ns_log notice "callback invoke search::index"
-                                callback -impl $driver search::index \
-                                    -object_id $object_id \
-                                    -content $txt \
-                                    -title $datasource(title) \
-                                    -keywords $datasource(keywords) \
-                                    -package_id $datasource(package_id) \
-                                    -community_id $datasource(community_id) \
-                                    -relevant_date $datasource(relevant_date) \
-                                    -datasource datasource
-                            } else {
-                                #ns_log notice "acs_sc::invoke FtsEngineDriver"
-                                set r [acs_sc::invoke \
-                                           -contract FtsEngineDriver \
-                                           -operation [expr {$event eq "UPDATE" ? "update_index" : "index"}] \
-                                           -call_args [list $datasource(object_id) \
-                                                           $txt $datasource(title) \
-                                                           $datasource(keywords)] \
-                                           -impl $driver]
-                            }
+			    foreach driver $drivers {
+				    ns_log notice "indexing with FtsEngineDriver $driver"
+				    if {[callback::impl_exists -callback search::index -impl $driver]} {
+					if {![info exists datasource(package_id)]} {
+					    set datasource(package_id) ""
+					}
+
+					if {![info exists datasource(relevant_date)]} {
+					    set datasource(relevant_date) ""
+					}
+					#ns_log notice "callback invoke search::index"
+					callback -impl $driver search::index \
+					    -object_id $object_id \
+					    -content $txt \
+					    -title $datasource(title) \
+					    -keywords $datasource(keywords) \
+					    -package_id $datasource(package_id) \
+					    -community_id $datasource(community_id) \
+					    -relevant_date $datasource(relevant_date) \
+					    -datasource datasource
+				    } else {
+					#ns_log notice "acs_sc::invoke FtsEngineDriver"
+					set r [acs_sc::invoke \
+						   -contract FtsEngineDriver \
+						   -operation [expr {$event eq "UPDATE" ? "update_index" : "index"}] \
+						   -call_args [list $datasource(object_id) \
+								   $txt $datasource(title) \
+								   $datasource(keywords)] \
+						   -impl $driver]
+				    }
+			    }
                         } errMsg]} {
                             ns_log Error "search::indexer: error getting datasource for " \
                                 "$object_id $object_type: $errMsg\n[ad_print_stack_trace]"
@@ -215,16 +232,21 @@
                 }
             }
             DELETE {
-                if {[catch {
-                    set r [acs_sc::invoke \
-                               -contract FtsEngineDriver \
-                               -operation unindex \
-                               -call_args [list $object_id] \
-                               -impl $driver]
-                } errMsg]} {
-                    ns_log Error "search::indexer: error unindexing $object_id " \
-                        "[acs_object_type $object_id]: $errMsg\n[ad_print_stack_trace]"
-                } else {
+		set ok 1
+		foreach driver $drivers {
+			if {[catch {
+			    set r [acs_sc::invoke \
+				       -contract FtsEngineDriver \
+				       -operation unindex \
+				       -call_args [list $object_id] \
+				       -impl $driver]
+			} errMsg]} {
+			    ns_log Error "search::indexer: error unindexing $object_id " \
+				"[acs_object_type $object_id]: $errMsg\n[ad_print_stack_trace]"
+			   set ok 0
+			}
+		}
+                if ($ok) {
                     # call the search action callbacks.
                     callback -catch search::action \
                         -action $event \
@@ -265,7 +287,7 @@
     storage_type
     object_id
 } {
-    @author Neophytos Demetriou
+    @author Neophytos Demetriou (neophytos@gmail.com)
 
     @param content
 
@@ -300,7 +322,7 @@
     _data
     mime
 } {
-    @author Neophytos Demetriou
+    @author Neophytos Demetriou (neophytos@gmail.com)
 } {
     upvar $_txt txt
     upvar $_data data
Index: www/search.adp
===================================================================
RCS file: /cvsroot/openacs-4/packages/search/www/search.adp,v
retrieving revision 1.19
diff -u -r1.19 search.adp
--- www/search.adp	7 Aug 2017 23:48:29 -0000	1.19
+++ www/search.adp	27 Jul 2023 09:15:37 -0000
@@ -58,6 +58,21 @@
   </div>
 </else>
 
+<if @vectorresult:rowcount@ ne 0>
+  <h3>Related Searches</h3>
+  <div id="related-searches">
+    <ul>
+      <multiple name="vectorresult">
+        <li>
+          <a href="@vectorresult.url_one@" class="vector-result-title">
+            @vectorresult.title;literal@
+          </a>
+	</li>
+      </multiple>
+    </ul>
+  </div>
+</if>
+
 <include src="/packages/search/lib/navbar" &="q"
     paginator_class="list-paginator-bottom" count="@result.count;literal@" &="low" &="high"
     &="offset" &="num" &="search_package_id">
Index: www/search.tcl
===================================================================
RCS file: /cvsroot/openacs-4/packages/search/www/search.tcl,v
retrieving revision 1.41.2.4
diff -u -r1.41.2.4 search.tcl
--- www/search.tcl	28 Feb 2022 20:14:25 -0000	1.41.2.4
+++ www/search.tcl	27 Jul 2023 09:15:37 -0000
@@ -1,5 +1,5 @@
 ad_page_contract {
-    @author Neophytos Demetriou <k2pts@cytanet.com.cy>
+    @author Neophytos Demetriou <neophytos@gmail.com>
     @creation-date September 01, 2001
     @cvs-id $Id: search.tcl,v 1.41.2.4 2022/02/28 20:14:25 gustafn Exp $
 } {
@@ -52,6 +52,7 @@
 
 set user_id [ad_conn user_id]
 set driver [parameter::get -package_id $package_id -parameter FtsEngineDriver]
+set vector_driver [parameter::get -package_id $package_id -parameter VectorDriver]
 if {[callback::impl_exists -impl $driver -callback search::driver_info]} {
     array set info [lindex [callback -impl $driver search::driver_info] 0]
     # array set info [list package_key intermedia-driver version 1 automatic_and_queries_p 1  stopwords_p 1]
@@ -92,7 +93,7 @@
 #set q [string tolower $q]
 set urlencoded_query [ad_urlencode $q]
 
-set params [list $q $offset $limit $user_id $df]
+set params [list $q $offset $limit $user_id $df $dt]
 if {$search_package_id eq ""
     && [parameter::get -package_id $package_id -parameter SubsiteSearchP -default 1]
     && [subsite::main_site_id] != [ad_conn subsite_id]
@@ -130,6 +131,12 @@
     array set result [acs_sc::invoke -contract FtsEngineDriver -operation search \
 			  -call_args $params -impl $driver]
 }
+
+ns_log notice vector_driver=$vector_driver params=$params
+array set vector_result [lindex [callback -impl $vector_driver search::search -query $q -offset 0 -limit 10 \
+				  -user_id $user_id -df $df \
+				  -extra_args [list package_ids $search_package_id object_type $object_type]] 0]
+
 set tend [clock clicks -milliseconds]
 
 if { $t eq [_ search.Feeling_Lucky] && $result(count) > 0} {
@@ -205,6 +212,28 @@
     }
 }
 
+template::multirow create vectorresult title url_one object_id
+foreach object_id $vector_result(ids) {
+    if {[catch {
+        set object_type [acs_object_type $object_id]
+        if {[callback::impl_exists -impl $object_type -callback search::datasource]} {
+            array set datasource [lindex [callback -impl $object_type search::datasource -object_id $object_id] 0]
+            set url_one [lindex [callback -impl $object_type search::url -object_id $object_id] 0]
+        } else {
+            #ns_log warning "SEARCH search/www/search.tcl callback::datasource::$object_type not found"
+            array set datasource [acs_sc::invoke -contract FtsContentProvider -operation datasource \
+				      -call_args [list $object_id] -impl $object_type]
+            set url_one [acs_sc::invoke -contract FtsContentProvider -operation url \
+			     -call_args [list $object_id] -impl $object_type]
+	    set title $datasource(title)
+        }
+    } errmsg]} {
+        ns_log error "search.tcl object_id $object_id object_type $object_type error $errmsg"
+    } else {
+        template::multirow append vectorresult $title $url_one
+    }
+}
+
 set search_the_web [parameter::get -package_id $package_id -parameter SearchTheWeb]
 if {$search_the_web ne ""} {
     set stw ""
