diff --git a/distfunc.c b/distfunc.c
index 853bba9..7f12c8d 100644
--- a/distfunc.c
+++ b/distfunc.c
@@ -11,10 +11,12 @@
 #define PORTABLE_ALIGN32 __declspec(align(32))
 #endif
 
+#define elts_per_vector sizeof(__m256) / sizeof(float)
+
 __attribute__((target("avx2,fma")))
 static dist_t l2_dist_impl_avx2(const coord_t *x, const coord_t *y, size_t n)
 {
-    const size_t elts_per_vector = sizeof(__m256) / sizeof(float);
+    // const size_t elts_per_vector = sizeof(__m256) / sizeof(float);
     const size_t unroll = 2;
     const size_t elts_per_loop = elts_per_vector * unroll;
     float partial_result[elts_per_vector];
